p2panda-engine/src/stream/ingest.rs

```diff
@@ -0,0 +1,258 @@
+    S: OperationStore<L, E> + LogStore<L, E>,
+    E: Clone + Serialize + DeserializeOwned + Extension<L> + Extension<PruneFlag>,
+{
+    pub(super) fn new(stream: St, store: S, ooo_buffer_size: usize) -> Ingest<St, S, L, E> {
+        let (ooo_buffer_tx, ooo_buffer_rx) = mpsc::channel::<IngestAttempt<E>>(ooo_buffer_size);
+
+        Ingest {
+            store,
+            stream: stream.fuse(),
+            ooo_buffer_size,
+            ooo_buffer_tx,
+            ooo_buffer_rx,
+            _marker: PhantomData,
+        }
+    }
+
+    delegate_access_inner!(stream, St, (.));
+}
+
+impl<St, S, L, E> Stream for Ingest<St, S, L, E>
+where
+    St: Stream<Item = (Header<E>, Option<Body>)>,
+    S: Clone + OperationStore<L, E> + LogStore<L, E>,
+    E: Clone + Serialize + DeserializeOwned + Extension<L> + Extension<PruneFlag>,
+{
+    type Item = Result<Operation<E>, IngestError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut store = self.store.clone();
+        let mut this = self.project();
+
+        loop {
+            // 1. Pull in the next item from the external stream or out-of-order buffer.
+            let res = {
+                // If the buffer ran full we prioritize pulling from it first, re-attempting
+                // ingest. This avoids clogging up the pipeline
+                if this.ooo_buffer_rx.size_hint().0 == *this.ooo_buffer_size {
+                    ready!(this.ooo_buffer_rx.as_mut().poll_next(cx))
+                } else {
+                    // Otherwise prefer pulling from the external stream first as freshly incoming
+                    // data should be prioritized
+                    match this.stream.as_mut().poll_next(cx) {
+                        Poll::Ready(Some((header, body))) => Some(IngestAttempt(header, body, 1)),
+                        Poll::Pending => ready!(this.ooo_buffer_rx.as_mut().poll_next(cx)),
+                        Poll::Ready(None) => match this.ooo_buffer_rx.as_mut().poll_next(cx) {
+                            Poll::Ready(Some(attempt)) => Some(attempt),
+                            // If there's no value coming from the buffer _and_ the external stream is
+                            // terminated, we can be sure nothing will come anymore
+                            Poll::Pending => None,
+                            Poll::Ready(None) => None,
+                        },
+                    }
+                }
+            };
+            let Some(IngestAttempt(header, body, counter)) = res else {
+                // Both external stream and buffer stream has ended, so we stop here as well
+                return Poll::Ready(None);
+            };
+
+            // 2. Validate and check the log-integrity of the incoming operation. If it is valid it
+            //    get's persisted and the log optionally pruned.
+            let ingest_fut = async { ingest_operation::<S, L, E>(&mut store, header, body).await };
```

Comment:
Aha, so if we allowed passing in log id as a parameter, then it would actually make it not possible to use generic stream patterns like this. Interesting! So our extensions are indeed very cool :wink: 

Am I right saying this? Had you considered this as well @adzialocha

Reply:
> Aha, so if we allowed passing in log id as a parameter, then it would actually make it not possible to use generic stream patterns like this. Interesting! So our extensions are indeed very cool 😉

Can you clarify what you mean with "parameter"? Do you mean an argument on the `ingest` method?

Reply:
It would be possible to not require the log id to be encoded in the header extension in the standalone `ingest_operation` method, which is cool, but yeah, for this stream we need to get it from each operation from somewhere. We can extract it before we call the method and make the extension mandatory when using streams.

Reply:
Thinking about it more: its possible to implement an extension trait without encoding anything. You could return a constant value instead in an extension impl, etc.